src/domain/datasets/services/src/services/key_blocks/dataset_key_block_update_handler.rs

```diff
@@ -39,21 +42,67 @@ use crate::collect_dataset_key_blocks_in_range;
 pub struct DatasetKeyBlockUpdateHandler {
     dataset_registry: Arc<dyn DatasetRegistry>,
     dataset_key_block_repo: Arc<dyn DatasetKeyBlockRepository>,
+    outbox: Arc<dyn Outbox>,
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 impl DatasetKeyBlockUpdateHandler {
+    async fn handle_dataset_reference_updated_message(
+        &self,
+        message: &DatasetReferenceMessageUpdated,
+    ) -> Result<(), InternalError> {
+        let target = match self
+            .dataset_registry
+            .get_dataset_by_id(&message.dataset_id)
+            .await
+        {
+            Ok(target) => Ok(target),
+            Err(odf::DatasetRefUnresolvedError::NotFound(e)) => {
+                tracing::error!(
+                    %message.dataset_id, err = ?e,
+                    "Updating dataset key blocks skipped. Dataset not found."
+                );
+                return Ok(());
+            }
+            Err(odf::DatasetRefUnresolvedError::Internal(e)) => Err(e),
+        }?;
+
+        let CollectKeyBlockResponse {
+            key_blocks,
+            divergence_detected: _, // We don't care about divergence here
+        } = self.update_dataset_key_blocks(target, message).await?;
```

Comment:
I'm not sure I understand how divergence is handled. But perhaps a valid behavior is to indicate it as a flag and allow to re-compute derived data from scratch. We can later optimize if divergence point is not that far from the end of the chain.